src/vault/client/http.clj

```diff
@@ -0,0 +1,496 @@
+        {:form-params {:app_id app, :user_id user}
+         :content-type :json
+         :accept :json
+         :as :json}))))
+
+
+
+;; ## Timer Logic
+
+(defn- try-renew-lease!
+  "Attempts to renew the given secret lease. Updates the lease store or catches
+  and logs any exception."
+  [client secret]
+  (try
+    (vault/renew-lease client (:lease-id secret))
+    (catch Exception ex
+      (log/error ex "Failed to renew secret lease" (:lease-id secret)))))
+
+
+(defn- try-rotate-secret!
+  "Attempts to rotate the given secret lease. Updates the lease store or catches
+  and logs any exception."
+  [client secret]
+  (try
+    (log/info "Rotating secret lease" (:lease-id secret))
+    (let [response (api-request client :get (:path secret) {})
+          info (assoc (clean-body response) :path (:path secret))]
+      (lease/update! (:leases client) info))
+    (catch Exception ex
+      (log/error ex "Failed to rotate secret" (:lease-id secret)))))
+
+
+(defn- maintain-leases!
+  [client window]
+  (log/debug "Checking for renewable leases...")
+  ; Check auth token for renewal.
+  (let [auth @(:auth client)]
+    (when (and (:renewable auth)
+               (lease/expires-within? auth window)
+               (some :lease-id (lease/list-leases (:leases client))))
+      (try
+        (log/info "Renewing Vault client token")
+        (vault/renew-token client)
+        (catch Exception ex
+          (log/error ex "Failed to renew client token!")))))
+  ; Renew leases that are within expiry window and are configured for renewal.
+  ; Rotate secrets that are about to expire and not renewable.
+  (let [renewable (lease/renewable-leases (:leases client) window)
+        rotatable (lease/rotatable-leases (:leases client) window)]
+    (doseq [secret renewable]
+      (try-renew-lease! client secret))
+    ; Rotate leases that are within expiry window and not renewable.
+    (doseq [secret rotatable]
+      (try-rotate-secret! client secret)))
+  ; Drop any expired leases.
+  (lease/sweep! (:leases client)))
+
+
+
+;; ## HTTP Client Type
+
+;; - `:api-url`
+;;   The base URL for the Vault API endpoint.
+;; - `:auth`
+;;   An atom containing the authentication lease information, including the
+;;   client token.
+;; - `:leases`
+;;   Local in-memory storage of secret leases.
+;; - `:lease-timer`
+;;   Thread which periodically checks and renews leased secrets.
+;; - `:lease-renewal-window
+;;   Period in seconds to renew leases before they expire.
+;; - `:lease-check-period`
+;;   Period in seconds to check for leases to renew.
+;; - `:lease-check-jitter`
+;;   Maximum amount in seconds to jitter the check period by.
+(defrecord HTTPClient
+  [api-url auth leases lease-timer]
+
+  component/Lifecycle
+
+  (start
+    [this]
+    (if lease-timer
+      ; Already running
+      this
+      ; Start lease heartbeat thread.
+      (let [window (:lease-renewal-window this 1200)  ; 20 minutes
```

Comment:
Nitpick: I prefer to do the math at runtime which self-documents the time units involved. e.g. `(* 20 60)`